### PR TITLE
[Theme] Support for screenshots

### PR DESCRIPTION
**Pull request by pamil**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | -
| License         | MIT

### CODE REVIEW

**Comment by adamelso** [comment]
Ah, the one thing that was missing from my April Fool's PR #4669 :+1:  . Speaking of, while working on it I couldn't see how to add the theme via the admin. I had to manually create a database record, then update the current channel to reference it.

File: src/Sylius/Bundle/ThemeBundle/Controller/ThemeScreenshotController.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        $this->themeRepository = $themeRepository;
+    }
+
+    /**
+     * @param int $themeId
+     * @param int $screenshotNumber
+     *
+     * @return BinaryFileResponse
+     */
+    public function streamScreenshotAction($themeId, $screenshotNumber)
```
Comment (aramalipoor) [inline src/Sylius/Bundle/ThemeBundle/Controller/ThemeScreenshotController.php]: What about `$screenshotCode`? e.g. `homepage`, `product`, `cart`, etc.

Comment (pamil) [inline src/Sylius/Bundle/ThemeBundle/Controller/ThemeScreenshotController.php]: The main issue here is that we can have both `screenshots/backend/product.jpg` and `screenshots/frontend/product.jpg` at the same time. `main.jpg` also is a valid screenshot path as it's the file in theme root directory

Comment (aramalipoor) [inline src/Sylius/Bundle/ThemeBundle/Controller/ThemeScreenshotController.php]: What about restricting the convention like:
```
/screenshots/main.jpg
/screenshots/backend-product.jpg
/screenshots/frontend-product.jpg
```
 * Only one-level directory
 * Code is the same as filename (+extension?)

File: src/Sylius/Bundle/ThemeBundle/Configuration/ThemeConfiguration.php [hunk 2]
```diff
@@ -89,6 +90,20 @@ private function addOptionalParentsList(ArrayNodeDefinition $rootNodeDefinition)
     /**
      * @param ArrayNodeDefinition $rootNodeDefinition
      */
+    private function addOptionalScreenshotsList(ArrayNodeDefinition $rootNodeDefinition)
+    {
+        $parentsNodeDefinition = $rootNodeDefinition->children()->arrayNode('screenshots');
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ThemeBundle/Configuration/ThemeConfiguration.php]: $screenshotsNodeDefinition

File: src/Sylius/Bundle/ThemeBundle/Tests/Configuration/ThemeConfigurationTest.php [hunk 3]
```diff
@@ -271,6 +271,76 @@ public function its_parents_replaces_other_parents_defined_elsewhere()
...
     }
 
+    /**
+     * @test
+     */
+    public function its_screenshots_are_strings()
+    {
+        $this->assertConfigurationIsValid(
+            [
+                ['screenshots' => ['screenshot/krzysztof-krawczyk.jpg', 'screenshot/ryszard-rynkowski.jpg']],
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ThemeBundle/Tests/Configuration/ThemeConfigurationTest.php]: 🎸

**Comment by pjedrzejewski** [comment]
Nice!

File: src/Sylius/Bundle/ThemeBundle/Controller/ThemeScreenshotController.php (hunk 1 above)
Comment (pamil) [inline src/Sylius/Bundle/ThemeBundle/Controller/ThemeScreenshotController.php]: I'm not sure whether we should enforce that convention, what do you think about referencing to the screenshots by a hash generated from their paths?

File: src/Sylius/Bundle/ThemeBundle/Configuration/ThemeConfiguration.php (hunk 2 above)
Comment (pamil) [inline src/Sylius/Bundle/ThemeBundle/Configuration/ThemeConfiguration.php]: I'm not too good at copypasting 😿

File: src/Sylius/Bundle/ThemeBundle/Controller/ThemeScreenshotController.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        $screenshotPath = $this->getScreenshotPath($this->getTheme($themeId), $screenshotNumber);
+
+        try {
+            return new BinaryFileResponse($screenshotPath);
+        } catch (FileNotFoundException $exception) {
+            throw new NotFoundHttpException(sprintf('Screenshot "%s" does not exist', $screenshotPath), $exception);
```
Comment (Zales0123) [inline src/Sylius/Bundle/ThemeBundle/Controller/ThemeScreenshotController.php]: Missing dot at the end of exception message :)

**Comment by pjedrzejewski** [comment]
Thanks Kamil!

File: src/Sylius/Bundle/ThemeBundle/Controller/ThemeScreenshotController.php (hunk 1 above)
Comment (aramalipoor) [inline src/Sylius/Bundle/ThemeBundle/Controller/ThemeScreenshotController.php]: You know what I was looking for in this was that we can somehow provide a developer-friendly way to ask for a screenshot. e.g. In themes tables lets show `homepage.jpg`. Hmmm, it needs more thinking I guess.

**Pull request merged**